lib/ansible/plugins/loader.py

```diff
@@ -1056,6 +1086,15 @@ def all(self, *args, **kwargs):
             else:
                 fqcn = f"ansible.builtin.{basename}"
             self._update_object(obj, basename, path, resolved=fqcn)
+            if self._plugin_instance_cache is not None:
+                context = PluginLoadContext()
+                context.resolved = True
+                context.original_name = basename
+                context.plugin_resolved_name = basename
+                context.plugin_resolved_path = path
+                context.plugin_resolved_collection = 'ansible.builtin' if fqcn.startswith('ansible.builtin.') else ''
+                context._resolved_fqcn = fqcn
+                self._plugin_instance_cache[path] = (obj, context)
```

Comment:
Distributing a copy/paste of this logic here doesn't feel right- `all()` is already duplicating a bunch of PluginLoader enumeration/instantiation logic as it is. I think you're on the right track with factoring out some of the `get_with_context` stuff into separate methods in this PR, so maybe `all` can piggyback on some of that to avoid duplicating the creation of the load context.

Reply:
I updated `all()` to call `get_with_context` to store the plugin in the cache and removed the external PluginLoader changes. The reason I made a separate cache using the path  as the key (as opposed to the plugin name) was to avoid the _resolve_step, which is slow. To work around that, `get_with_context` now can short-circuit before `_resolve_step` is called (for vars plugins only). I left a code comment about why that's there, in case we expand the cache to handle other plugin types, since that could be an issue for plugin types that use the collections list and need to be resolved every time.